File: 프로그래머스/LV2/코딩테스트 연습/월간 코드 챌린지 시즌3/빛의 경로 사이클/빛의 경로 사이클.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#define d pair<int, int>

using namespace std;

vector<vector<set<d>>> d_record;
vector<d> directions = { d(-1,0), d(0,1), d(1,0), d(0,-1) };
vector<int> path_lens;
d start_d, start_rc;
int R, C, directions_len = directions.size();

void go(int r, int c, int di, d start_rc, d start_d, int path_len, vector<string>& grid)
{
    if (grid[r][c] == 'L')di = ((di + 1) + directions_len) % directions_len;
    else if (grid[r][c] == 'R')di = ((di - 1) + directions_len) % directions_len;

    if (d_record[r][c].find(directions[di]) != d_record[r][c].end())return;
    else d_record[r][c].insert(directions[di]);

    if (d(r, c) == start_rc && directions[di] == start_d)
    {
        path_lens.push_back(path_len);
        return;
    }

    int next_r = ((r + directions[di].second) + R) % R;
    int next_c = ((c + directions[di].first) + C) % C;
    go(next_r, next_c, di, start_rc, start_d, path_len + 1, grid);
}

vector<int> solution(vector<string> grid)
{
    vector<set<d>> temp;
    R = grid.size();
    C = grid[0].length();
    int path_len = 1;

    for (int r = 0; r < R; r++)
    {
        temp.clear();
        for (int c = 0; c < C; c++)
        {
            set<d> s;
            temp.push_back(s);
        }
        d_record.push_back(temp);
    }

    for (int r = 0; r < R; r++)
    {
        for (int c = 0; c < C; c++)
        {
            for (int di = 0; di < directions_len; di++)
            {
                if (d_record[r][c].find(directions[di]) != d_record[r][c].end())continue;
                start_rc = d(r, c);
                start_d = d(directions[di].first, directions[di].second);
                int next_r = ((r + directions[di].second) + R) % R;
                int next_c = ((c + directions[di].first) + C) % C;
                go(next_r, next_c, di, start_rc, start_d, path_len, grid);
            }
        }
    }
    sort(path_lens.begin(), path_lens.end());
    return path_lens;
}

File: 프로그래머스/LV2/코딩테스트 연습/월간 코드 챌린지 시즌3/빛의 경로 사이클/빛의 경로 사이클.py
import sys
sys.setrecursionlimit(1500000)

def solution(grid):
    d_record = []
    path_lens = []
    row, col = len(grid), len(grid[0])
    directions = [[0,-1],[1,0],[0,1],[-1,0]]
    directions_len = len(directions)
    pathlen = 1

    for r in range(row):
        temp = []
        for c in range(col):
            temp.append(set())
        d_record.append(temp)
    
    def go(r, c, di, start, pathlen):
        if grid[r][c] == 'L':
            di = ((di+1)+directions_len) % directions_len
        elif grid[r][c] == 'R':
            di = ((di-1)+directions_len) % directions_len
            
        new_d = directions[di]

        newd_str = "".join(map(str,new_d))
        if newd_str in d_record[r][c]: return
        else: d_record[r][c].add(newd_str)

        if [r,c, new_d] == start:
            path_lens.append(pathlen)
            return
        
        next_r = ((r + new_d[1])+row) % row
        next_c = ((c + new_d[0])+col) % col
        go(next_r, next_c, di, start, pathlen+1)

    for r in range(row):
        for c in range(col):
            for di in range(len(directions)):
                d_str = "".join(map(str,directions[di]))
                if d_str in d_record[r][c]:
                    continue
                start = [r,c,directions[di]]
                next_r = ((r + directions[di][1])+row) % row
                next_c = ((c + directions[di][0])+col) % col
                go(next_r, next_c, di, start, pathlen)
    path_lens.sort()
    return path_lens